--- mini_httpd-1.19.orig/mini_httpd.c	2009-11-02 10:25:31.734520575 +0200
+++ mini_httpd-1.19/mini_httpd.c	2009-11-02 10:29:19.582021172 +0200
@@ -232,9 +232,11 @@
 
 /* Forwards. */
 static void usage( void );
+#ifdef USE_CONFIG
 static void read_config( char* filename );
 static void value_required( char* name, char* value );
 static void no_value_required( char* name, char* value );
+#endif
 static int initialize_listen_socket( usockaddr* usaP );
 static void handle_request( void );
 static void de_dotdot( char* file );
@@ -351,11 +353,13 @@
 	    (void) printf( "%s\n", SERVER_SOFTWARE );
 	    exit( 0 );
 	    }
+	#ifdef USE_CONFIG
 	else if ( strcmp( argv[argn], "-C" ) == 0 && argn + 1 < argc )
 	    {
 	    ++argn;
 	    read_config( argv[argn] );
 	    }
+	#endif
 	else if ( strcmp( argv[argn], "-D" ) == 0 )
 	    debug = 1;
 #ifdef USE_SSL
@@ -554,7 +558,7 @@
 		*/
 	{
 	    //printf("\ncertfile=%s\n", certfile);
-	    
+
 	    if (ssl_obj_load (ssl_ctx, SSL_OBJ_RSA_KEY, certfile, 0)!=SSL_OK)
 	    {
 		//printf("ssl_obj_load() error");
@@ -861,19 +865,22 @@
 	}
     }
 
-
+  
 static void
 usage( void )
     {
+	(void) fprintf( stderr, "usage:  %s "
 #ifdef USE_SSL
-    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-S] [-E certfile] [-Y cipher] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
-#else /* USE_SSL */
-    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
-#endif /* USE_SSL */
+	"[-S] [-E certfile] [-Y cipher] "
+#endif
+#ifdef USE_CONFIG
+    "[-C configfile] "
+#endif
+    "-V [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
     exit( 1 );
     }
 
-
+#ifdef USE_CONFIG
 static void
 read_config( char* filename )
     {
@@ -1063,6 +1070,7 @@
 	exit( 1 );
 	}
     }
+#endif
 
 
 static int
@@ -3074,7 +3082,7 @@
 
     /* Just set a flag that we got the signal. */
     got_hup = 1;
-	
+
     /* Restore previous errno. */
     errno = oerrno;
     }
