diff -rNu mini_httpd-1.19.orig/Makefile mini_httpd-1.19/Makefile
--- mini_httpd-1.19.orig/Makefile	2002-11-02 01:02:57.000000000 +0200
+++ mini_httpd-1.19/Makefile	2009-04-13 12:37:06.478019133 +0300
@@ -14,17 +14,19 @@
 # http://www.openssl.org/  Make sure the SSL_TREE definition points to the
 # tree with your OpenSSL installation - depending on how you installed it,
 # it may be in /usr/local instead of /usr/local/ssl.
-#SSL_TREE =	/usr/local/ssl
-#SSL_DEFS =	-DUSE_SSL
-#SSL_INC =	-I${SSL_TREE}/include
-#SSL_LIBS =	-L${SSL_TREE}/lib -lssl -lcrypto
-
+SSL_TREE ?=	../axTLS
+SSL_DEFS =	-DUSE_SSL
+SSL_INC =	-I${SSL_TREE}/ssl -I${SSL_TREE}/config -I${SSL_TREE}/crypto
+
+#SSL_LIBS =	-L${SSL_TREE}/_stage -laxtls -lcrypto
+SSL_LIBS =	-L${SSL_TREE}/_stage -laxtls
+#SSL_LIBS =	-L${SSL_TREE}/_stage -lcrypt ${SSL_TREE}/_stage/libaxtls.a
 
 BINDIR =	/usr/local/sbin
 MANDIR =	/usr/local/man
-CC =		gcc
+#CC =		/usr/local/arm/3.4.3-0.9.27/arm-linux/bin/gcc
 CDEFS =		${SSL_DEFS} ${SSL_INC}
-CFLAGS =	-O ${CDEFS}
+CFLAGS +=	-O ${CDEFS}
 #CFLAGS =	-g ${CDEFS}
 LDFLAGS =	-s
 #LDFLAGS =	-g
diff -rNu mini_httpd-1.19.orig/keys/certs/ca.cer mini_httpd-1.19/keys/certs/ca.cer
--- mini_httpd-1.19.orig/keys/certs/ca.cer	1970-01-01 02:00:00.000000000 +0200
+++ mini_httpd-1.19/keys/certs/ca.cer	2009-01-16 11:37:18.000000000 +0200
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
+MIIDjTCCAvagAwIBAgIJAJ8YAGNTJCrsMA0GCSqGSIb3DQEBBQUAMIGMMQswCQYD
+VQQGEwJSTzESMBAGA1UECBMJQnVjaGFyZXN0MRIwEAYDVQQHEwlCdWNoYXJlc3Qx
+EDAOBgNVBAoTB0NDb3N0aW4xEDAOBgNVBAsTB0NDb3N0aW4xDzANBgNVBAMTBkNv
+c3RpbjEgMB4GCSqGSIb3DQEJARYRY2Nvc3RpbkBnbWFpbC5jb20wHhcNMDkwMTE2
+MDkzNzE4WhcNMTExMDE0MDkzNzE4WjCBjDELMAkGA1UEBhMCUk8xEjAQBgNVBAgT
+CUJ1Y2hhcmVzdDESMBAGA1UEBxMJQnVjaGFyZXN0MRAwDgYDVQQKEwdDQ29zdGlu
+MRAwDgYDVQQLEwdDQ29zdGluMQ8wDQYDVQQDEwZDb3N0aW4xIDAeBgkqhkiG9w0B
+CQEWEWNjb3N0aW5AZ21haWwuY29tMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKB
+gQChtLjwLlII4jF7F5QzfzoMqSTG9CICw81wJaxFVu3ydEA3R5vv7+OJt+RSAxl1
+tmxVuNYcg2D4f2zt4yTZrVjL4JleMyCnQw0WYkDK+58yO98ZxI1fMmd0KL8Ja6Wx
+XZ6qx0nmBmPF5Ii3XdercZ9GfxrBoVCByOAx+EYvKmALBwIDAQABo4H0MIHxMB0G
+A1UdDgQWBBT0n1VYQlrfwYtmdw5guf2FtoEQajCBwQYDVR0jBIG5MIG2gBT0n1VY
+QlrfwYtmdw5guf2FtoEQaqGBkqSBjzCBjDELMAkGA1UEBhMCUk8xEjAQBgNVBAgT
+CUJ1Y2hhcmVzdDESMBAGA1UEBxMJQnVjaGFyZXN0MRAwDgYDVQQKEwdDQ29zdGlu
+MRAwDgYDVQQLEwdDQ29zdGluMQ8wDQYDVQQDEwZDb3N0aW4xIDAeBgkqhkiG9w0B
+CQEWEWNjb3N0aW5AZ21haWwuY29tggkAnxgAY1MkKuwwDAYDVR0TBAUwAwEB/zAN
+BgkqhkiG9w0BAQUFAAOBgQBKzy6H87gkeLeRcVGDe0iN0Lal0nGAUVvSnOxUcPoN
+EPau84fWGXC92NF7VH5GqG0JHumMSWp0NCALosn5abytDeSW5ssD/evCg0UZyztM
+Bozn3rJeHWUefwIauzqsbsvedT3D/DNQqaU0giEbpP8WfdC8VgGsjAaIVS9vvvJi
+Pw==
+-----END CERTIFICATE-----
diff -rNu mini_httpd-1.19.orig/keys/keys/ca.key mini_httpd-1.19/keys/keys/ca.key
--- mini_httpd-1.19.orig/keys/keys/ca.key	1970-01-01 02:00:00.000000000 +0200
+++ mini_httpd-1.19/keys/keys/ca.key	2009-01-16 11:35:08.000000000 +0200
@@ -0,0 +1,18 @@
+-----BEGIN RSA PRIVATE KEY-----
+Proc-Type: 4,ENCRYPTED
+DEK-Info: DES-EDE3-CBC,6FD9671CBE1E2DFB
+
+d3oF10Rzqb95794XtTu1EEsUXvkAkZqbY06d00OgOM6eh/sDR008QN2i4Vn/wqRF
+KWLoy7RrqfpS3r0zs3ZvA3Gq9Gvo9mFNLjDHFM6A0cQOYhdRcyvGxjrj8H5Igrm6
+lcFk1OXUA10D6xjVBk7p4U3pEES+IGpOAw6/vEzsIWsxOjmgBHd5HU4xko8+yHNI
+tqhokEaI+xWq6dNA8W9aSJ6o/b3ubpnaVFACeSUTPi0wd1xkAIpR5INo/CxH7VPr
+Y2GwNUU46ibD7q9QpUkNgU2Mce3dy+w8Exbtwa83+fnPvciOT/hb4RlV4c5+th/v
+S+isJchlaqSYKFq/TDYKGYIDbW5ZjH4XAkXL3IUyEhD68J+XnChrSGuK9m4wEGrs
+3O6dexzpnlhZ66F8uB7NJgLVoANux4zWxn8n7+RHmEAfUqljVaYi1++C3nXbmZt4
+UuJxC4S3sTsvxDOEBxvgb0j+gdpaw59F1xhYejD1E2JhjcNFI+NsjtFO0+ejoEpR
+cAN5N0Zh403qn/vkjrFp//j+RVPAAhdvixaDYMW41LKqSx1g5zIP72t1bgXLgY3X
+Urw1+jB9YwYAZDK4jBNkb+TVdczHyKHWAmfAd2D60Y+tP5VLVHb48ozaf4YO+5Kt
+G9MczwVAeO+g4GTqkrxVbCF3QE/PXMOT3OKlrKtWltbZBZgb7IbOPfH1DlCf3JsZ
+35EHzilebalm8JhkU6VyB3ZTmATZeqTUpt3Oc7EIntkl0yvUfg7jw+593EBVPmCr
+mjiDpeJ6yohylfzbIfo+xT3E5jDWGeTdygEBXIycaQ8ph49cWbUkeA==
+-----END RSA PRIVATE KEY-----
diff -rNu mini_httpd-1.19.orig/mini_httpd.c mini_httpd-1.19/mini_httpd.c
--- mini_httpd-1.19.orig/mini_httpd.c	2005-06-29 20:31:17.000000000 +0300
+++ mini_httpd-1.19/mini_httpd.c	2009-04-13 16:04:04.386383740 +0300
@@ -1,6 +1,6 @@
 /* mini_httpd - small HTTP server
 **
-** Copyright © 1999,2000 by Jef Poskanzer <jef@mail.acme.com>.
+** Copyright Â© 1999,2000 by Jef Poskanzer <jef@mail.acme.com>.
 ** All rights reserved.
 **
 ** Redistribution and use in source and binary forms, with or without
@@ -52,6 +52,7 @@
 #include <arpa/inet.h>
 #include <netdb.h>
 #include <dirent.h>
+#include <grp.h>
 
 #include "port.h"
 #include "match.h"
@@ -66,8 +67,7 @@
 #endif /* HAVE_SENDFILE */
 
 #ifdef USE_SSL
-#include <openssl/ssl.h>
-#include <openssl/err.h>
+#include "ssl.h"
 #endif /* USE_SSL */
 
 extern char* crypt( const char* key, const char* setting );
@@ -540,17 +540,28 @@
 #ifdef USE_SSL
     if ( do_ssl )
 	{
-	SSL_load_error_strings();
-	SSLeay_add_ssl_algorithms();
+//	SSL_load_error_strings();
+	//SSLeay_add_ssl_algorithms();
 	ssl_ctx = SSL_CTX_new( SSLv23_server_method() );
 	if ( certfile[0] != '\0' )
-	    if ( SSL_CTX_use_certificate_file( ssl_ctx, certfile, SSL_FILETYPE_PEM ) == 0 ||
+	    /*if ( SSL_CTX_use_certificate_file( ssl_ctx, certfile, SSL_FILETYPE_PEM ) == 0 ||
 		 SSL_CTX_use_PrivateKey_file( ssl_ctx, certfile, SSL_FILETYPE_PEM ) == 0 ||
 		 SSL_CTX_check_private_key( ssl_ctx ) == 0 )
 		{
 		ERR_print_errors_fp( stderr );
 		exit( 1 );
 		}
+		*/
+	{
+	    //printf("\ncertfile=%s\n", certfile);
+	    
+	    if (ssl_obj_load (ssl_ctx, SSL_OBJ_RSA_KEY, certfile, 0)!=SSL_OK)
+	    {
+		//printf("ssl_obj_load() error");
+		ERR_print_errors_fp( stderr );
+		exit(1);
+	    }
+	}
 	if ( cipher != (char*) 0 )
 	    {
 	    if ( SSL_CTX_set_cipher_list( ssl_ctx, cipher ) == 0 )
@@ -608,7 +619,7 @@
 	    syslog( LOG_CRIT, "%s - %m", pidfile );
 	    perror( pidfile );
             exit( 1 );
-            }
+             }
         (void) fprintf( pidfp, "%d\n", (int) getpid() );
         (void) fclose( pidfp );
         }
@@ -1363,6 +1374,7 @@
 
 #ifdef USE_SSL
     SSL_free( ssl );
+    SSL_CTX_free(ssl_ctx);
 #endif /* USE_SSL */
     }
 
@@ -2347,6 +2359,7 @@
 
 #ifdef USE_SSL
     SSL_free( ssl );
+    SSL_CTX_free(ssl_ctx);
 #endif /* USE_SSL */
     exit( 1 );
     }
@@ -2634,7 +2647,10 @@
     {
 #ifdef USE_SSL
     if ( do_ssl )
-	return SSL_read( ssl, buf, size );
+    {
+    	//printf("SSL_READ\n");
+		return SSL_read( ssl, buf, size );
+	}
     else
 	return read( conn_fd, buf, size );
 #else /* USE_SSL */
@@ -2648,7 +2664,10 @@
     {
 #ifdef USE_SSL
     if ( do_ssl )
-	return SSL_write( ssl, buf, size );
+    {
+    	//printf("SSL_WRITE\n");
+		return SSL_write( ssl, buf, size );
+	}
     else
 	return write( conn_fd, buf, size );
 #else /* USE_SSL */
