--- mini_httpd-1.19.orig/mini_httpd.c	2009-11-02 10:47:54.798020804 +0200
+++ mini_httpd-1.19/mini_httpd.c	2009-11-02 11:00:35.262520617 +0200
@@ -189,6 +189,7 @@
 static char* charset;
 static char* p3p;
 static int max_age;
+static int read_timeout;
 static FILE* logfp;
 static int listen4_fd, listen6_fd;
 #ifdef USE_SSL
@@ -338,6 +339,7 @@
     charset = DEFAULT_CHARSET;
     p3p = (char*) 0;
     max_age = -1;
+    read_timeout = READ_TIMEOUT ;
     user = DEFAULT_USER;
     hostname = (char*) 0;
     logfile = (char*) 0;
@@ -438,6 +440,11 @@
 	    ++argn;
 	    max_age = atoi( argv[argn] );
 	    }
+	else if ( strcmp( argv[argn], "-R" ) == 0 && argn + 1 < argc )
+	    {
+	    ++argn;
+	    read_timeout = atoi( argv[argn] );
+	    }
 	else
 	    usage();
 	++argn;
@@ -868,7 +875,7 @@
 	}
     }
 
-  
+
 static void
 usage( void )
     {
@@ -879,7 +886,7 @@
 #ifdef USE_CONFIG
     "[-C configfile] "
 #endif
-    "-V [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
+    "-V [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage] [-R readtimeout]\n", argv0 );
     exit( 1 );
     }
 
